### ISSUE TITLE
Loaders - to load or not to load

### ISSUE THREAD

**Issue by Deadlock989** [Deadlock989/IndustrialRevolution#240]
Progress so far, representing about 12 hours work:

![image](https://github.com/Deadlock989/IndustrialRevolution/assets/3128845/fe6663a0-9f5e-44e5-8f26-38f586e35af1)

Context: recently one of the Factorio devs decided to implement powered loaders, and also have them interact with train wagons. We had been asking for both of these things for years. There is currently a game-crashing issue with steam-powered loaders which is apparently fixed for the next experimental release of Factorio. Burner loaders are not possible because inserters would place fuel on the belt.

In terms of graphics and specifically render layer issues, fluid-powered loaders represent the worst possible case imaginable, with electric or free loaders and underground belts coming in a close second and third. These issues were multiplied when belts became "fat" and had an elevation off the ground (see [FFF#269](https://www.factorio.com/blog/post/fff-269)), requiring undergrounds to get a new layered structure (loaders inherit a similar structure). There are still many (rare) render layer issue cases with underground belts in vanilla which the devs have marked as "won't fix" (inserter arms are another serial offender). For fluid powered loaders, you have the entity object layer, the belt layer, the item layer and the shadow layer all trying to interact in the same 64x64 pixel grid, in different ways depending on the facing direction, with pipes forcing you to overstep grid boundaries (pipe are above belts and items), plus the weirdness of the integration patch layer rotating with loader input/output mode - and all in the context of Factorio's weird perspective (which belts violate). The way that render layers are implemented in Factorio means that in this case it is impossible to have absolutely correct-looking shadows, and there is only one way to have a loader that guarantees no render layer issues: make it into a literally square box (e.g. Krastorio / Vanilla Loaders), usually with weird perspective distortion that does cover the belt from the isometric perspective but doesn't look like it's the right shape or even physically possible (Exotic Industries / Angel's / Loader Redux). 

The best attempt so far is AAI Loaders which also had to tackle the fluid pipe connection issue: it does a good job but there are still problems if you define "not being pixel perfect" as a problem. My attempt fixes a couple of those issues but at the cost of introducing new ones, so, eh.

I'm going to sit on this until the fluid power fix for loaders makes it into a stable release. Option 1 is to forget about this - adding AAI Loaders works well if you want loaders in IR3 and they are visually the best on the market so far. Option 2 is to put my loaders into IR3 directly and make them opt-in, glitchy warts and all. Option 3 is to spend another bunch of hours trying to kill more render layer edge cases, add some ingot stacking and a "new beltbox" that fits with the loader's geometry better than IR3 small assemblers do, and make it separate mod.

**Comment by Deadlock989** [comment]
Got rid of 3/4 of the shadow issues by some scripted rendering hacks. There remain some odd render layer situations (mostly shadows) when unlikely combinations of entities are put near to each other but there are several cases of this kind of thing in vanilla so I'm not going to lose any sleep over it.

Have put the loaders into a separate mod with a kind of "beltbox stacking lite" attachment - ingots, bricks, bars and sheets can be stacked, nothing else. 

![image](https://github.com/Deadlock989/IndustrialRevolution/assets/3128845/567f0a7e-73ec-4dfa-86ed-ecbb7af46c06)

**Comment by Bilka2** [comment]
> some scripted rendering hacks. 

Bystander question: Does this mean runtime rendering, so LuaRendering (`rendering.*`)? If yes, I'd advise against heavy use of it. While the performance of script rendering was improved, it is quite far away from the heavy optimisations of entity rendering. This is mostly relevant when using high numbers of render objects, such as attaching them to tens of thousands of entities of a certain type. Replacing (stationary) entity graphics is not the primary usecase of script rendering, so it is optimised for something else (ease of moving/changing targets).

[To sum up](https://forums.factorio.com/viewtopic.php?p=502371#p502371):
> Script rendering system doesn't have any spacial organization so during rendering every object is checked for validity and if it collides with viewport, so that's the FPS impact.

**Comment by Deadlock989** [comment]
Yes, runtime rendering. Thank you for the reminder that there is a consequence to it. 

In this picture only one of the loaders has a rendered patch, bottom left:

![image](https://github.com/Deadlock989/IndustrialRevolution/assets/3128845/1d74cc56-d7e4-4202-a5bb-869d1a18e924)

- The patches are only needed on east-west facing loaders, they are never used for north-south ones
- Patches are removed when another belt-type entity is placed to the south

So how many of your loaders get patched depends on how you build. 

I just ran a test, placing 8700 patched loaders into the world. The "game render preparation" time stat did increase hugely, by about 2.5x. There was no impact on FPS (and my rig is well out of date). But it is concerning. That said, 8000+ loaders is a lot (given that not all loaders need a patch, you will probably have many times more than that even).

I could just set an upper limit on the number of patches - e.g. 5000 - so that the minor ugliness starts to manifest only once you've reached a huge scale and are unlikely to be spending much time inspecting belts at maximum zoom anyhow.

Or I could use a forceless non-colliding entity that is amenable to some level of render layer manipulation - e.g. a SimpleEntity. It would have to be cleaned up in the same way as currently, and that's bound to create issues, the benefit of rendered sprites is that you can attach them to an entity as a target and then not have to worry about manually removing them if the entity is deleted.

**Comment by Deadlock989** [comment]
Just like old times.

![stacking2023](https://github.com/Deadlock989/IndustrialRevolution/assets/3128845/2aa0ea75-b5b6-4e17-9a89-04954218fbf5)

This will be released after IR 3.1.0 which will be after Factorio next updates its stable release branch.
